include/ck/utility/functional3.hpp

```diff
@@ -86,31 +86,121 @@ struct ford_impl<Sequence<>, Orders>
+                              IDim0 * IDim1 * IDim2 * IDim3,
+                              IDim0 * IDim1 * IDim2 * IDim3 * IDim4>;
+};
+
+template <int32_t IDim0, int32_t IDim1, int32_t IDim2, int32_t IDim3, int32_t IDim4, int32_t IDim5>
+struct make_cumulative_product<IDim0, IDim1, IDim2, IDim3, IDim4, IDim5>
 {
-    __host__ __device__ constexpr static_ford()
+    using type = ck::Sequence<IDim0,
+                              IDim0 * IDim1,
+                              IDim0 * IDim1 * IDim2,
+                              IDim0 * IDim1 * IDim2 * IDim3,
+                              IDim0 * IDim1 * IDim2 * IDim3 * IDim4,
+                              IDim0 * IDim1 * IDim2 * IDim3 * IDim4 * IDim5>;
+};
+// clang-format on
+template <int32_t... Dims>
+struct convert_flat_to_multi_index
+{
+    using SDim                 = ck::Sequence<Dims...>;
+    static constexpr auto Prod = (Dims * ...);
+    using TCumProd             = typename make_cumulative_product<Dims...>::type;
+
+    template <ck::index_t flat_idx>
+    using type = decltype((TCumProd{} * ck::Number<flat_idx>{} / ck::Number<Prod>{}) % SDim{});
+};
+
+template <typename T, T... Is>
+struct range_applier
+{
+    // F: code block parameterized by compile-time constant
+    // IndexTransform: metafunction from int32_t to code block argument type
+    // Result: side effect of executing the code block for each element in (Is...)
+    template <typename F, template <int32_t> typename IndexTransform = ck::Number>
+    __host__ __device__ constexpr void operator()(F f) const
     {
-        static_assert(Lengths::GetSize() > 0, "wrong! Lengths is empty");
-        static_assert(Lengths::GetSize() == Orders::GetSize(), "wrong! inconsistent size");
+        static_assert(sizeof...(Is) <= 3136, "tweak -fbracket-depth");
+        (f(IndexTransform<Is>{}), ...);
     }
+};
 
-    // F signature: F(Sequence<...> multi_id)
-    // multi_id is the unordered multi-index
-    template <class F>
+} // namespace detail
+
+// Lengths is Sequence<...>, it is the length of each dimension for
+// N-dimensional loop
+template <typename T>
+struct static_ford;
+
+template <template <int32_t...> typename T, int32_t... Dims>
+struct static_ford<T<Dims...>>
+    : __make_integer_seq<detail::range_applier, ck::index_t, (Dims * ...)>
+{
+    // `base` is the same as `range_applier<index_t, 0, ..., product of Dims>`
+    using base = __make_integer_seq<detail::range_applier, ck::index_t, (Dims * ...)>;
+
+    template <ck::index_t I>
+    using convert_t = typename detail::convert_flat_to_multi_index<Dims...>::template type<I>;
+
+    template <typename F, template <auto> typename IndexTransform = convert_t>
     __host__ __device__ constexpr void operator()(F f) const
     {
-        constexpr auto ordered_lengths = Lengths::ReorderGivenNew2Old(Orders{});
-        detail::static_ford_impl<decltype(ordered_lengths), Orders>{}(f, Sequence<>{});
+        base::template operator()<F, IndexTransform>(f);
```

Comment:
possible improvement here is to figure out how to bake index transform into the base class, to do `using base::operator()` and avoid one extra template instantiation